File: 백준/Gold/15732. 도토리 숨기기/도토리 숨기기.py
import sys

input = sys.stdin.readline

N, K, D = map(int, input().split())

rules = [list(map(int, input().split())) for _ in range(K)]

left = 1
right = N

while left <= right:
    mid = (left + right) // 2

    cnt = 0
    for start, end, step in rules:
        if start > mid:
            continue

        if mid > end:
            cnt += (end - start) // step + 1

        else:
            cnt += (mid - start) // step + 1

    if cnt >= D:
        answer = mid
        right = mid - 1
    else:
        left = mid + 1

print(answer)    

File: 백준/Gold/15732. 도토리 숨기기/도토리 숨기기.java
import java.io.*;
import java.util.*;

public class Main {
    
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int K = Integer.parseInt(st.nextToken());
        int D = Integer.parseInt(st.nextToken());

        List<int[]> rules = new ArrayList<>(); // 0: start 1: end 2: dis

        for (int i = 0; i < K; i++) {
            st = new StringTokenizer(br.readLine());
            rules.add(new int[] {Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())});
        }

        int left = 0;
        int right = N + 1;

        while (left + 1 < right) {
            int mid = (left + right) / 2;

            long cornsToPut = calculate(mid, rules);

            if (cornsToPut < D) {
                left = mid;
            } else if (cornsToPut >= D) {
                right = mid;
            }
        }

        System.out.println(right);
    }

    private static long calculate(int mid, List<int[]> rules) {
        long tmp = 0;

        for (int[] r : rules) {
            int start = r[0];
            int end = r[1];
            int dis = r[2];

            if (mid < start) continue;
            else if (mid > end) {
                tmp += (end - start) / dis + 1;
            } else {
                tmp += (mid - start) / dis + 1;
            }
        }

        return tmp;
    }
}
